qiskit/circuit/library/hamiltonian_gate.py

```diff
@@ -0,0 +1,142 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Gate described by the time evolution of a Hermitian Hamiltonian operator.
+"""
+
+from __future__ import annotations
+import typing
+
+from numbers import Number
+import numpy as np
+
+from qiskit.circuit import Gate, QuantumCircuit, QuantumRegister, ParameterExpression
+from qiskit.circuit.exceptions import CircuitError
+from qiskit.quantum_info.operators.predicates import matrix_equal
+from qiskit.quantum_info.operators.predicates import is_hermitian_matrix
+from qiskit.utils.deprecation import deprecate_func
+
+from .generalized_gates.unitary import UnitaryGate
+
+if typing.TYPE_CHECKING:
+    from qiskit.quantum_info import BaseOperator
+
+
+class HamiltonianGate(Gate):
+    r"""Class for representing evolution by a Hamiltonian operator as a gate.
+
+    This gate resolves to a :class:`.UnitaryGate` as :math:`U(t) = \exp(-i t H)`,
+    which can be decomposed into basis gates if it is 2 qubits or less, or
+    simulated directly in Aer for more qubits.
+    """
+
+    def __init__(
+        self,
+        data: np.ndarray | Gate | BaseOperator,
+        time: float | ParameterExpression,
+        label: str | None = None,
+    ) -> None:
+        """
+        Args:
+            data: A hermitian operator.
+            time: Time evolution parameter.
+            label: Unitary name for backend [Default: None].
+
+        Raises:
+            ValueError: if input data is not an N-qubit unitary operator.
+        """
+        if hasattr(data, "to_matrix"):
+            # If input is Gate subclass or some other class object that has
+            # a to_matrix method this will call that method.
+            data = data.to_matrix()
+        elif hasattr(data, "to_operator"):
+            # If input is a BaseOperator subclass this attempts to convert
+            # the object to an Operator so that we can extract the underlying
+            # numpy matrix from `Operator.data`.
+            data = data.to_operator().data
+        # Convert to np array in case not already an array
+        data = np.array(data, dtype=complex)
+        # Check input is unitary
+        if not is_hermitian_matrix(data):
+            raise ValueError("Input matrix is not Hermitian.")
```

Comment:
Some errors were changed from `ExtensionsError` to `ValueError`, which is strictly speaking not backwards compatible. It seemed small enough to be still fine, but we could also create an intermediate class, covering both, for the deprecation period.